### DISCUSSION TITLE
Misaligned logical date when dagrun time exceeds schedule_interval

### DISCUSSION THREAD

**Discussion by NickYadance** [apache/airflow#25393] [General] [answered]
Airflow Version: `2.3.2`

The DAG definition:
```
with DAG(
    dag_id=dag_id,
    default_args=default_args,
    description=TASK_NAME,
    catchup=False,
    concurrency=20,
    max_active_runs=4,
    schedule_interval=timedelta(days=1),
    start_date=pendulum.datetime(2022, 6, 27, 0, 30, 0, tz="Asia/Singapore")
) as dag:
```
The dagrun list:
<img width="1714" alt="image" src="https://user-images.githubusercontent.com/10060849/181727591-ba7cdb4e-4655-48f3-8fc9-8e031121e3e9.png">
The expected logical_date is :

`dag start date + n * schedule_interval`

Seems that the real logical_date is (when last dagrun duration exceeds schedule_interval)

`last end date - scheduler_interval`

Is this the expected behavior for logical_date ?

**Comment by potiuk** [comment]
Good question @uranusjr ?

**Comment by NickYadance** [comment]
I also test this with setting `schedule_interval` to cron expression and `max_active_runs=2`:
```
from datetime import timedelta

import pendulum
from airflow import DAG
from airflow.operators.bash import BashOperator

with DAG(
        '__test',
        default_args={
            'depends_on_past': False,
            'email_on_failure': True,
            'email_on_retry': False,
            'retries': 0,
            'retry_delay': timedelta(minutes=1),
        },
        description='For internal use',
        schedule_interval='* * * * *',
        start_date=pendulum.datetime(2022, 5, 23, tz='Asia/Singapore'),
        max_active_runs=2,
        catchup=False,
        tags=['example'],
) as dag:
    sleep = BashOperator(task_id='sleep',
                         bash_command='sleep')
    sleep
```

The result surprises me too as some logical date is missing if the previous dagruns run for too long.
<img width="1272" alt="image" src="https://user-images.githubusercontent.com/10060849/182568575-fba333b8-34a1-4d33-a9de-6a1d29b92c43.png">

**Reply by NickYadance** [reply]
I tried looking for the reason for this in scheduler and FileProcessor of Airflow however failed to find any. Maybe somewhere else is setting `next_dagrun` for the DAG except for FileProcessor ?

**Comment by uranusjr** [comment]
The logical date has nothing to do with the end date, it’s only calculated from the schedule. The second comment identifies the issue correctly—if the previous run is still running when the schedule reaches the next trigger, that trigger is simply skipped. But I believe this is the expected behaviour since you set `catchup=False`.

https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html#catchup

**Comment by NickYadance** [comment]
https://github.com/apache/airflow/blob/298be502c35006b7c3f011b676dbb4db0633bc74/airflow/timetables/interval.py#L351-L362
https://github.com/apache/airflow/blob/298be502c35006b7c3f011b676dbb4db0633bc74/airflow/timetables/interval.py#L221-L242

Both `DeltaDataIntervalTimetable` and `CronDataIntervalTimetable ` will infer `new_start` from now and this matches the scheduling behavior observed from the test. 

And most likely the catchup is used to catch up data from the `start_date` of the DAG , to the time the first dagrun is triggered, which is namely **catch up history data**. but what you are saying is **catch up history and ongoing data**, which is a little confusing to me and hard to realize.

Maybe I'd better write a custom timetable that is **consistent** to achieve consistent scheduling dagruns.

**Reply by uranusjr** [reply]
It’s a bit of confusing naming, but _start_ and _end_ in the code you referenced and the UI mean different things. In the timetable, they refer to the run’s _data interval_ (which is an interval and thus has a start and an end), while in the UI they mean when the DAG run starts running and when it ends running.

**Reply by uranusjr** [reply]
But regardless, this is how `catchup` works. When a run is “missed” because the previous run has not finished, the data it represents become history data, and is therefore considered no longer relevant if you set `catchup=False`. Since this definition does not work for you, what do you feel Airflow should distinguish “history data”?

**Reply by NickYadance** [reply]
I'll close this for now since the issue is resloved, thx for your support.

**Comment by NickYadance** [comment, marked as answer]
To anyone running into the same issue, the solution is to to turn on `catchUp=True` for you DAG.
